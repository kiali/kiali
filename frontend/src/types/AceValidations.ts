import { HelpMessage, ObjectCheck, ObjectValidation } from './IstioObjects';
import { Annotation } from 'react-ace/types';
import { IMarker } from 'react-ace';
import {YAMLException} from "js-yaml";

export const jsYaml = require('js-yaml');

export interface AceValidations {
  markers: Array<IMarker>;
  annotations: Array<Annotation>;
}

interface AceCheck {
  marker: IMarker;
  annotation: Annotation;
}

interface AceMarker {
  startRow: number;
  startCol: number;
  endRow: number;
  endCol: number;
  position: number;
}

interface YamlPosition {
  position: number;
  row: number;
  col: number;
}

const numRows = (yaml: string): number => {
  let rows = 0;
  for (let i = 0; i < yaml.length; i++) {
    if (yaml.charAt(i) === '\n') {
      rows++;
    }
  }
  return rows;
};

const posToRowCol = (yaml: string, pos: number): YamlPosition => {
  const rowCol: YamlPosition = {
    position: pos,
    row: 0,
    col: 0
  };
  let lastNL = -1;
  for (let i = 0; i < pos; i++) {
    if (yaml.charAt(i) === '\n') {
      rowCol.row++;
      lastNL = i;
    }
  }
  rowCol.col = lastNL > -1 ? pos - (lastNL + 1) : pos;
  return rowCol;
};

export const rowColToPos = (yaml: string, row: number, col: number): number => {
  let currentRow = 0;
  let currentCol = 0;
  const pos = -1;
  for (let i = 0; i < yaml.length; i++) {
    if (yaml.charAt(i) === '\n') {
      currentRow++;
      currentCol = -1;
    } else {
      currentCol++;
    }
    if (currentRow === row && currentCol === col) {
      // If col == 0, pos is NL char, so returned pos should be first char after NL
      return col === 0 ? i + 1 : i;
    }
  }
  return pos;
};

export const parseLine = (yaml: string, row: number): string => {
  let i = 0;
  let j = 0;

  for (i; i < yaml.length; i++) {
    if (yaml.charAt(i) === '\n') {
      j = j + 1;
    }

    if (j === row) break;
  }

  return yaml.substring(i + 1, yaml.indexOf('\n', i + 1));
};

export const parseHelpAnnotations = (yaml: string, helpMessages: HelpMessage[]): Annotation[] => {
  let annotations: Annotation[] = [];
  let lastPosition = -1;

  helpMessages.forEach(hm => {
    const marker = parseMarker(
      yaml,
      lastPosition,
      hm.objectField.substring(hm.objectField.lastIndexOf('.') + 1),
      false
    );

    const annotation = {
      row: marker.startRow,
      column: marker.startCol,
      type: 'info',
      text: 'This field has help information. Check the side panel for more information.'
    };

    if (marker.position !== -1) {
      annotations.push(annotation);
    }
  });

  return annotations;
};

/*
  Find a token inside a yaml based string.
  Returns the row/col coordinates of the token.
  It manages special cases where a token is an array.
 */
const parseMarker = (
  yaml: string,
  startsFrom: number,
  token: string,
  isArray: boolean,
  arrayIndex?: number
): AceMarker => {
  const aceMarker: AceMarker = {
    startRow: 0,
    startCol: 0,
    endRow: 0,
    endCol: 0,
    position: -1
  };

  let tokenPos = startsFrom;

  // Find start of the spec part first, this should skip the whole metadata part
  if (startsFrom < 0) {
    tokenPos = yaml.indexOf('spec:', tokenPos);
  }

  // Find initial token position
  tokenPos = yaml.indexOf(token, tokenPos);
  if (tokenPos < 0) {
    return aceMarker;
  }

  const maxRows = numRows(yaml);

  // Array should find first '-' token to situate pos
  if (isArray && arrayIndex !== undefined) {
    tokenPos = yaml.indexOf('-', tokenPos);
    // We should find the right '-' under the same col of the yaml
    const firstArrayRowCol = posToRowCol(yaml, tokenPos);
    let row = firstArrayRowCol.row;
    const col = firstArrayRowCol.col;
    let arrayIndexPos = tokenPos;
    let indexRow = 0;
    // Iterate to find next '-' token according arrayIndex
    while (row < maxRows && indexRow < arrayIndex) {
      row++;
      const checkPos = rowColToPos(yaml, row, col);
      if (yaml.charAt(checkPos) === '-') {
        arrayIndexPos = checkPos;
        indexRow++;
      }
    }
    const arrayRowCol = posToRowCol(yaml, arrayIndexPos);
    aceMarker.position = arrayIndexPos + 1; // Increase the index to not repeat same finding on next iteration
    aceMarker.startRow = arrayRowCol.row;
    aceMarker.startCol = arrayRowCol.col;
  } else {
    const tokenRowCol = posToRowCol(yaml, tokenPos);
    aceMarker.position = tokenPos + token.length; // Increase the index to not repeat same finding on next iteration
    aceMarker.startRow = tokenRowCol.row;
    aceMarker.startCol = tokenRowCol.col;
  }

  // Once start is calculated, we should calculate the end of the element iterating by rows
  for (let row = aceMarker.startRow + 1; row < maxRows + 1; row++) {
    // It searches by row and column, starting from the beginning of the line
    for (let col = 0; col <= aceMarker.startCol; col++) {
      const endTokenPos = rowColToPos(yaml, row, col);
      // We need to differentiate if token is an array or not to mark the end of the mark
      if (yaml.charAt(endTokenPos) !== ' ' && (isArray || yaml.charAt(endTokenPos) !== '-')) {
        aceMarker.endRow = row;
        aceMarker.endCol = 0;
        return aceMarker;
      }
    }
  }
  return aceMarker;
};

const parseCheck = (yaml: string, check: ObjectCheck): AceCheck => {
  const severity = check.severity === 'error' || check.severity === 'warning' ? check.severity : 'info';
  const marker: IMarker = {
    startRow: 0,
    startCol: 0,
    endRow: 0,
    endCol: 0,
    className: 'istio-validation-' + severity,
    type: 'fullLine'
  };
  const annotation = {
    row: 0,
    column: 0,
    type: severity,
    text: (check.code ? check.code + ' ' : '') + check.message
  };
  let aceMarker = {
    startRow: 0,
    startCol: 0,
    endRow: 0,
    endCol: 0,
    position: -1
  };
  /*
    Potential paths:
      - <empty, no path>
      - spec/hosts
      - spec/host
      - spec/<protocol: http|tcp>[<nRoute>]/route
      - spec/<protocol: http|tcp>[<nRoute>]/route[nDestination]
      - spec/<protocol: http|tcp>[<nRoute>]/route[<nDestination>]/weight/<value>
      - spec/<protocol: http|tcp>[nRoute]/route[nDestination]/destination
   */
  if (check.path.length > 0) {
    const tokens: string[] = check.path.split('/');
    // It skips the first 'spec' token
    if (tokens.length > 1) {
      for (let i = 1; i < tokens.length; i++) {
        const token = tokens[i];
        // Check if token has an array or not
        if (token.indexOf('[') > -1 && token.indexOf(']') > -1) {
          const startPos = token.indexOf('[');
          const endPos = token.indexOf(']');
          const arrayIndex = +token.substr(startPos + 1, endPos - startPos - 1);
          const subtoken = token.substr(0, startPos);
          aceMarker = parseMarker(yaml, aceMarker.position, subtoken, true, arrayIndex);
        } else {
          aceMarker = parseMarker(yaml, aceMarker.position, token, false);
        }
      }
    }
  }

  marker.startRow = aceMarker.startRow;
  marker.startCol = aceMarker.startCol;
  // React Ace editor has a flip in the marker indexes
  marker.endRow = aceMarker.endRow > 0 ? aceMarker.endRow - 1 : 0;
  marker.endCol = aceMarker.endCol;
  annotation.row = marker.startRow;
  return { marker: marker, annotation: annotation };
};

export const parseKialiValidations = (yamlInput: string, kialiValidations?: ObjectValidation): AceValidations => {
  const aceValidations: AceValidations = {
    markers: [],
    annotations: []
  };

  if (!kialiValidations || yamlInput.length === 0 || Object.keys(kialiValidations).length === 0) {
    return aceValidations;
  }

  kialiValidations.checks.forEach(check => {
    const aceCheck = parseCheck(yamlInput, check);
    aceValidations.markers.push(aceCheck.marker);
    aceValidations.annotations.push(aceCheck.annotation);
  });
  return aceValidations;
};

export const parseYamlValidations = (yamlInput: string): AceValidations => {
  const parsedValidations: AceValidations = {
    markers: [],
    annotations: []
  };
  try {
    jsYaml.safeLoadAll(yamlInput);
  } catch (e) {
    if (e instanceof YAMLException) {
      const row = e.mark && e.mark.line ? e.mark.line : 0;
      const col = e.mark && e.mark.column ? e.mark.column : 0;
      const message = e.message ? e.message : '';
      parsedValidations.markers.push({
        startRow: row,
        startCol: 0,
        endRow: row + 1,
        endCol: 0,
        className: 'istio-validation-error',
        type: 'fullLine'
      });
      parsedValidations.annotations.push({
        row: row,
        column: col,
        type: 'error',
        text: message
      });
    }
  }
  return parsedValidations;
};
