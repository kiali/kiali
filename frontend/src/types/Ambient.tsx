export enum WaypointType {
  All = 'all',
  None = 'none',
  Service = 'service',
  Workload = 'workload'
}

export const Waypoint = 'waypoint';
export const WaypointForLabel = 'istio.io/waypoint-for';

export const AmbientComponent = 'ambient';
