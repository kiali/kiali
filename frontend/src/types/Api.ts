import axios, { AxiosResponse, AxiosError } from 'axios';

export interface KialiError {
  detail: string;
  error: string;
}

export type ApiResponse<T> = Partial<AxiosResponse<T>> & {
  data: T;
};

export type ApiError = AxiosError<KialiError>;

export const isApiError = axios.isAxiosError;
