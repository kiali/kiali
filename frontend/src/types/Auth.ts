export interface AuthConfig {
  authorizationEndpoint?: string;
  logoutEndpoint?: string;
  logoutRedirect?: string;
  strategy: AuthStrategy;
}

export type AuthInfo = {
  sessionInfo: SessionInfo;
} & AuthConfig;

export enum AuthStrategy {
  anonymous = 'anonymous',
  openshift = 'openshift',
  token = 'token',
  openid = 'openid',
  header = 'header'
}

// Stores the result of a computation:
// hold = stop all computation and wait for a side-effect, such as a redirect
// continue = continue...
// success = authentication was a success, session is available
// failure = authentication failed, session is undefined but error is available
export enum AuthResult {
  HOLD = 'hold',
  CONTINUE = 'continue',
  SUCCESS = 'success',
  FAILURE = 'failure'
}

export interface SessionInfo {
  username?: string;
  expiresOn?: string;
}

export const getCSRFToken = (): string | undefined => {
  const cookiePrefix = 'csrf-token=';

  return (
    document &&
    document.cookie &&
    document.cookie
      .split(';')
      .map(c => c.trim())
      .filter(c => c.startsWith(cookiePrefix))
      .map(c => c.slice(cookiePrefix.length))
      .pop()
  );
};
