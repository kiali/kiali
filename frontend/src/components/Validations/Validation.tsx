import * as React from 'react';
import {
  CheckCircleIcon,
  ExclamationCircleIcon,
  ExclamationTriangleIcon,
  InfoCircleIcon
} from '@patternfly/react-icons';
import { ValidationTypes } from '../../types/IstioObjects';
import { Text, TextVariants } from '@patternfly/react-core';
import { PFColors } from 'components/Pf/PfColors';
import { kialiStyle } from 'styles/StyleUtils';
import { IconProps, createIcon } from 'config/KialiIcon';

const validationStyle = kialiStyle({
  textAlign: 'left',
  $nest: {
    '&:last-child p': {
      margin: 0
    }
  }
});

type ValidationProps = {
  message?: string;
  messageColor?: boolean;
  severity: ValidationTypes;
};

const ErrorValidation: IconProps = {
  color: PFColors.Danger,
  icon: ExclamationCircleIcon,
  dataTest: 'icon-error-validation'
};

const WarningValidation: IconProps = {
  color: PFColors.Warning,
  icon: ExclamationTriangleIcon,
  dataTest: 'icon-warning-validation'
};

const InfoValidation: IconProps = {
  color: PFColors.Info,
  icon: InfoCircleIcon
};

const CorrectValidation: IconProps = {
  color: PFColors.Success,
  icon: CheckCircleIcon,
  dataTest: 'icon-correct-validation'
};

const severityToValidation: { [severity: string]: IconProps } = {
  correct: CorrectValidation,
  error: ErrorValidation,
  info: InfoValidation,
  warning: WarningValidation
};

export const Validation: React.FC<ValidationProps> = (props: ValidationProps) => {
  const validation = severityToValidation[props.severity];
  const severityColor = { color: validation.color };
  const hasMessage = !!props.message;

  // Set styles
  const textStyle = props.messageColor ? severityColor : {};
  const iconStyle = kialiStyle(severityColor);

  const iconProps: IconProps = {
    className: iconStyle,
    icon: validation.icon,
    dataTest: validation.dataTest
  };

  if (hasMessage) {
    return (
      <div className={validationStyle}>
        <Text component={TextVariants.p} style={textStyle}>
          {createIcon(iconProps)} {props.message}
        </Text>
      </div>
    );
  } else {
    return <>{createIcon(iconProps)}</>;
  }
};
