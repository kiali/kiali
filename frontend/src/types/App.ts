import { Namespace } from './Namespace';
import { Runtime, WaypointInfo } from './Workload';
import { InstanceType } from 'types/Common';
import { AppHealthResponse } from '../types/Health';
import { GroupVersionKind } from './IstioObjects';

export type AppId = {
  app: string;
  cluster?: string;
  namespace: string;
};

export interface AppWorkload {
  ambient?: string;
  isAmbient: boolean;
  isGateway: boolean;
  istioSidecar: boolean;
  labels: { [key: string]: string };
  namespace: string;
  serviceAccountNames: string[];
  waypointWorkloads?: WaypointInfo[];
  workloadGVK: GroupVersionKind;
  workloadName: string;
}

export interface App {
  cluster?: string;
  health: AppHealthResponse;
  instanceType: InstanceType.App;
  isAmbient: boolean;
  name: string;
  namespace: Namespace;
  runtimes: Runtime[];
  serviceNames: string[];
  workloads: AppWorkload[];
}

export interface AppQuery {
  health: 'true' | 'false';
  rateInterval: string;
}
