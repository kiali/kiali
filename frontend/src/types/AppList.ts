import { AppHealth } from './Health';
import { ObjectReference } from './IstioObjects';
import { InstanceType } from 'types/Common';

export interface AppList {
  applications: AppListItem[];
  cluster?: string;
}

export interface AppListItem {
  ambient?: string;
  cluster?: string;
  health: AppHealth;
  instanceType: InstanceType.App;
  isAmbient: boolean;
  isGateway: boolean;
  istioReferences: ObjectReference[];
  istioSidecar: boolean;
  labels: { [key: string]: string };
  name: string;
  namespace: string;
}

export interface AppListQuery {
  health: 'true' | 'false';
  istioResources: 'true' | 'false';
  rateInterval: string;
}
