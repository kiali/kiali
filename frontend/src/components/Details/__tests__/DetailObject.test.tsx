import * as React from 'react';
import { shallow } from 'enzyme';
import { default as DetailObject } from '../DetailObject';
import { HTTPRouteDestination, ValidationTypes } from '../../../types/IstioObjects';
import { shallowToJson } from 'enzyme-to-json';
import Validation from '../../Validations/Validation';

describe('DetailObject test', () => {
  const detail: HTTPRouteDestination = {
    destination: {
      host: 'reviews',
      subset: 'v1',
      port: {
        number: 22,
        name: 'ssh'
      }
    },
    weight: 85
  };

  const mockRandom = () => {
    const mockMath = Object.create(global.Math);
    mockMath.random = () => 0.8;
    global.Math = mockMath;
  };

  it('prints a nested list with all attributes in the detail', () => {
    mockRandom();

    const wrapper = shallow(<DetailObject name={"name"} detail={detail} />);

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();

    expect(wrapper.html()).toContain('<span>[host]</span>');
    expect(wrapper.html()).toContain('<span>[subset]</span>');
    expect(wrapper.html()).toContain('<span>[weight]</span>');

    expect(wrapper.html()).toContain('<strong>port</strong>');
    expect(wrapper.html()).toContain('<span>[number]</span>');
    expect(wrapper.html()).toContain('<span>[name]</span>');
  });

  it("doesn't print excluded fields", () => {
    mockRandom();

    const wrapper = shallow(<DetailObject name={"name"} detail={detail} exclude={['port']} />);

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();

    expect(wrapper.html()).toContain('<span>[host]</span>');
    expect(wrapper.html()).toContain('<span>[subset]</span>');
    expect(wrapper.html()).toContain('<span>[weight]</span>');

    expect(wrapper.html()).not.toContain('<strong>port</strong>');
    expect(wrapper.html()).not.toContain('<span>[number]</span>');
    expect(wrapper.html()).not.toContain('<span>[name]</span>');
  });

  it('prints an alert message', () => {
    const validation = {
      message: 'Not all checks passed',
      severity: ValidationTypes.Error
    };

    mockRandom();

    const wrapper = shallow(<DetailObject name={"name"} detail={detail} validation={validation} />);

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();

    const iconWrapper = wrapper.find(Validation);
    expect(iconWrapper.prop('severity')).toEqual(validation.severity);
    expect(iconWrapper.prop('message')).toEqual(validation.message);
    expect(iconWrapper.prop('messageColor')).toEqual(true);
  });

  it("doesn't print any alert message", () => {
    const validation = {
      message: '',
      severity: ValidationTypes.Error
    };

    mockRandom();

    const wrapper = shallow(<DetailObject name={"name"} detail={detail} validation={validation} />);

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();

    const iconWrapper = wrapper.find(Validation);
    expect(iconWrapper.length).toEqual(0);
  });
});
