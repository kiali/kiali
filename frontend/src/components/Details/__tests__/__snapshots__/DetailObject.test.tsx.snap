// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailObject test doesn't print any alert message 1`] = `
<div>
  <div>
    <strong>
      name
    </strong>
    <ul
      className="details"
    >
      <li
        key="key_sssssssss_k0"
      >
        <div>
          <strong>
            destination
          </strong>
          <ul
            className="details"
          >
            <li
              key="key_sssssssss_k0"
            >
              <div>
                <span>
                  <span>
                    [host]
                  </span>
                   
                  reviews
                </span>
              </div>
            </li>
            <li
              key="key_sssssssss_k1"
            >
              <div>
                <span>
                  <span>
                    [subset]
                  </span>
                   
                  v1
                </span>
              </div>
            </li>
            <li
              key="key_sssssssss_k2"
            >
              <div>
                <strong>
                  port
                </strong>
                <ul
                  className="details"
                >
                  <li
                    key="key_sssssssss_k0"
                  >
                    <div>
                      <span>
                        <span>
                          [number]
                        </span>
                         
                        22
                      </span>
                    </div>
                  </li>
                  <li
                    key="key_sssssssss_k1"
                  >
                    <div>
                      <span>
                        <span>
                          [name]
                        </span>
                         
                        ssh
                      </span>
                    </div>
                  </li>
                </ul>
              </div>
            </li>
          </ul>
        </div>
      </li>
      <li
        key="key_sssssssss_k1"
      >
        <div>
          <span>
            <span>
              [weight]
            </span>
             
            85
          </span>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`DetailObject test doesn't print excluded fields 1`] = `
<div>
  <div>
    <strong>
      name
    </strong>
    <ul
      className="details"
    >
      <li
        key="key_sssssssss_k0"
      >
        <div>
          <strong>
            destination
          </strong>
          <ul
            className="details"
          >
            <li
              key="key_sssssssss_k0"
            >
              <div>
                <span>
                  <span>
                    [host]
                  </span>
                   
                  reviews
                </span>
              </div>
            </li>
            <li
              key="key_sssssssss_k1"
            >
              <div>
                <span>
                  <span>
                    [subset]
                  </span>
                   
                  v1
                </span>
              </div>
            </li>
            <li
              key="key_sssssssss_k2"
            />
          </ul>
        </div>
      </li>
      <li
        key="key_sssssssss_k1"
      >
        <div>
          <span>
            <span>
              [weight]
            </span>
             
            85
          </span>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`DetailObject test prints a nested list with all attributes in the detail 1`] = `
<div>
  <div>
    <strong>
      name
    </strong>
    <ul
      className="details"
    >
      <li
        key="key_sssssssss_k0"
      >
        <div>
          <strong>
            destination
          </strong>
          <ul
            className="details"
          >
            <li
              key="key_sssssssss_k0"
            >
              <div>
                <span>
                  <span>
                    [host]
                  </span>
                   
                  reviews
                </span>
              </div>
            </li>
            <li
              key="key_sssssssss_k1"
            >
              <div>
                <span>
                  <span>
                    [subset]
                  </span>
                   
                  v1
                </span>
              </div>
            </li>
            <li
              key="key_sssssssss_k2"
            >
              <div>
                <strong>
                  port
                </strong>
                <ul
                  className="details"
                >
                  <li
                    key="key_sssssssss_k0"
                  >
                    <div>
                      <span>
                        <span>
                          [number]
                        </span>
                         
                        22
                      </span>
                    </div>
                  </li>
                  <li
                    key="key_sssssssss_k1"
                  >
                    <div>
                      <span>
                        <span>
                          [name]
                        </span>
                         
                        ssh
                      </span>
                    </div>
                  </li>
                </ul>
              </div>
            </li>
          </ul>
        </div>
      </li>
      <li
        key="key_sssssssss_k1"
      >
        <div>
          <span>
            <span>
              [weight]
            </span>
             
            85
          </span>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`DetailObject test prints an alert message 1`] = `
<div>
  <div>
    <strong>
      name
    </strong>
    <Validation
      message="Not all checks passed"
      messageColor={true}
      severity="error"
    />
    <ul
      className="details"
    >
      <li
        key="key_sssssssss_k0"
      >
        <div>
          <strong>
            destination
          </strong>
          <ul
            className="details"
          >
            <li
              key="key_sssssssss_k0"
            >
              <div>
                <span>
                  <span>
                    [host]
                  </span>
                   
                  reviews
                </span>
              </div>
            </li>
            <li
              key="key_sssssssss_k1"
            >
              <div>
                <span>
                  <span>
                    [subset]
                  </span>
                   
                  v1
                </span>
              </div>
            </li>
            <li
              key="key_sssssssss_k2"
            >
              <div>
                <strong>
                  port
                </strong>
                <ul
                  className="details"
                >
                  <li
                    key="key_sssssssss_k0"
                  >
                    <div>
                      <span>
                        <span>
                          [number]
                        </span>
                         
                        22
                      </span>
                    </div>
                  </li>
                  <li
                    key="key_sssssssss_k1"
                  >
                    <div>
                      <span>
                        <span>
                          [name]
                        </span>
                         
                        ssh
                      </span>
                    </div>
                  </li>
                </ul>
              </div>
            </li>
          </ul>
        </div>
      </li>
      <li
        key="key_sssssssss_k1"
      >
        <div>
          <span>
            <span>
              [weight]
            </span>
             
            85
          </span>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
